confgenerator/confmerger.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package confgenerator provides functions to generate subagents configuration from unified agent.
+package confgenerator
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/spf13/viper"
+)
+
+var (
+	builtInConf = map[string]string{
+		"linux": `logging:
+  receivers:
+    syslog:
+      type: files
+      include_paths:
+      - /var/log/messages
+      - /var/log/syslog
+  exporters:
+    google:
+      type: google_cloud_logging
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [syslog]
+        exporters: [google]
+
+metrics:
+  receivers:
+    hostmetrics:
+      type: hostmetrics
+      collection_interval: 60s
+  exporters:
+    google:
+      type: google_cloud_monitoring
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [hostmetrics]
+        exporters: [google]`,
+
+		"windows": `logging:
+  receivers:
+    windows_event_log:
+      type: windows_event_log
+      channels: [System,Application,Security]
+  exporters:
+    google:
+      type: google_cloud_logging
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [windows_event_log]
+        exporters: [google]
+
+metrics:
+  receivers:
+    hostmetrics:
+      type: hostmetrics
+      collection_interval: 60s
+    mssql:
+      type: mssql
+      collection_interval: 60s
+    iis:
+      type: iis
+      collection_interval: 60s
+  exporters:
+    google:
+      type: google_cloud_monitoring
+  service:
+    pipelines:
+      default_pipeline:
+        receivers: [hostmetrics,mssql,iis]
+        exporters: [google]`,
+	}
+)
+
+func MergeConfFiles(builtInConfPath, userConfPath, mergedConfPath, platform string) error {
+	builtInConfFolder := path.Dir(builtInConfPath)
+	builtInConfFullName := path.Base(builtInConfPath)
+	builtInConfName := strings.TrimSuffix(builtInConfFullName, filepath.Ext(builtInConfFullName))
+	userConfFolder := path.Dir(userConfPath)
+	userConfFullName := path.Base(userConfPath)
+	userConfName := strings.TrimSuffix(userConfFullName, filepath.Ext(userConfFullName))
+	mergedConfFolder := path.Dir(mergedConfPath)
+
+	// Write the built-in conf to disk for debugging purpose.
+	f, err := os.Create(builtInConfPath)
```

Comment:
Does Create handle the situation when the directory doesn't exist? (Could that ever happen?)

Reply:
Creation of the directory is one of the issues that I'm still fixing. (Sent this PR out to get early feedback, as it could be a bit controversial)

Reply:
This has been handled now.